Upstream-Status: Inappropriate

Signed-off-by: Raman Rakavets <radikot88@gmail.com>
diff --git a/arch/arm/include/asm/spl.h b/arch/arm/include/asm/spl.h
index ee79a19c05c..cd9b7942fca 100644
--- a/arch/arm/include/asm/spl.h
+++ b/arch/arm/include/asm/spl.h
@@ -20,6 +20,7 @@ enum {
 	BOOT_DEVICE_NAND,
 	BOOT_DEVICE_ONENAND,
 	BOOT_DEVICE_NOR,
+	BOOT_DEVICE_SPINAND,
 	BOOT_DEVICE_UART,
 	BOOT_DEVICE_SPI,
 	BOOT_DEVICE_USB,
diff --git a/arch/arm/mach-rockchip/spl-boot-order.c b/arch/arm/mach-rockchip/spl-boot-order.c
index 6572dde29f6..aa8977963e9 100644
--- a/arch/arm/mach-rockchip/spl-boot-order.c
+++ b/arch/arm/mach-rockchip/spl-boot-order.c
@@ -76,6 +76,9 @@ static int spl_node_to_boot_device(int node)
 	if (!uclass_find_device_by_of_offset(UCLASS_SPI_FLASH, node, &parent))
 		return BOOT_DEVICE_SPI;
 
+	if (!uclass_get_device_by_of_offset(UCLASS_MTD, node, &parent))
+		return BOOT_DEVICE_SPINAND;
+
 	return -1;
 }
 
@@ -176,7 +179,7 @@ int spl_decode_boot_device(u32 boot_device, char *buf, size_t buflen)
 #endif
 	int ret;
 
-	if (boot_device == BOOT_DEVICE_SPI) {
+	if (boot_device == BOOT_DEVICE_SPI || boot_device == BOOT_DEVICE_SPINAND) {
 		/* Revert spl_node_to_boot_device() logic to find appropriate SPI flash device */
 
 		/*
@@ -188,6 +191,7 @@ int spl_decode_boot_device(u32 boot_device, char *buf, size_t buflen)
 		int elem;
 		int node;
 		const char *conf;
+		enum uclass_id u_id;
 
 		if (chosen_node < 0) {
 			debug("%s: /chosen not found\n", __func__);
@@ -219,7 +223,18 @@ int spl_decode_boot_device(u32 boot_device, char *buf, size_t buflen)
 				continue;
 			}
 
-			ret = uclass_find_device_by_of_offset(UCLASS_SPI_FLASH, node, &dev);
+			switch (boot_device) {
+			case BOOT_DEVICE_SPI:
+				u_id = UCLASS_SPI_FLASH;
+				break;
+			case BOOT_DEVICE_SPINAND:
+				u_id = UCLASS_MTD;
+				break;
+			default:
+				return -ENODEV;
+			}
+
+			ret = uclass_find_device_by_of_offset(u_id, node, &dev);
 			if (ret) {
 				debug("%s: could not find udevice for %s\n", __func__, conf);
 				continue;
diff --git a/common/spl/Kconfig b/common/spl/Kconfig
index 3b7b6cafef8..abe0fa213ed 100644
--- a/common/spl/Kconfig
+++ b/common/spl/Kconfig
@@ -968,6 +968,20 @@ config SPL_MTD
 	  devices. See SPL_NAND_SUPPORT and SPL_ONENAND_SUPPORT for how
 	  to enable specific MTD drivers.
 
+config SPL_MTD_LOAD
+	bool "Support loading from MTD framework"
+	depends on SPL_MTD
+	help
+	  Enable support for loading next stage, U-Boot or otherwise, from
+	  MTD Framework in U-Boot SPL.
+
+config SYS_MTD_U_BOOT_OFFS
+	hex "address of u-boot payload in MTD flash"
+	default 0x0
+	depends on SPL_MTD_LOAD
+	help
+	 Address within MTD Flash from where the u-boot payload is fetched.
+
 config SPL_MUSB_NEW
 	bool "Support new Mentor Graphics USB"
 	help
@@ -1044,6 +1058,12 @@ config SPL_RELOC_LOADER
 	  to temporary memory, then copying it into place afterwards, then
 	  jumping to it.
 
+config SPL_SPINAND_SUPPORT
+	bool "Support loading from SPI NAND flash"
+	depends on SPL_MTD_LOAD
+	help
+	  Enable support for SPI NAND Flash in SPL.
+
 config SPL_UBI
 	bool "Support UBI"
 	help
diff --git a/common/spl/Makefile b/common/spl/Makefile
index 4c9482bd309..9f5ddcad17a 100644
--- a/common/spl/Makefile
+++ b/common/spl/Makefile
@@ -35,6 +35,7 @@ obj-$(CONFIG_$(PHASE_)NVME) += spl_nvme.o
 obj-$(CONFIG_$(PHASE_)SEMIHOSTING) += spl_semihosting.o
 obj-$(CONFIG_$(PHASE_)DFU) += spl_dfu.o
 obj-$(CONFIG_$(PHASE_)SPI_LOAD) += spl_spi.o
+obj-$(CONFIG_$(PHASE_)MTD_LOAD) += spl_mtd.o
 obj-$(CONFIG_$(PHASE_)RAM_SUPPORT) += spl_ram.o
 obj-$(CONFIG_$(PHASE_)USB_SDP_SUPPORT) += spl_sdp.o
 endif
diff --git a/common/spl/spl_mtd.c b/common/spl/spl_mtd.c
new file mode 100644
index 00000000000..caa4f2fd503
--- /dev/null
+++ b/common/spl/spl_mtd.c
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) 2025 Raman Rakavets <radikot88@gmail.com>
+ */
+ 
+#include <config.h>
+#include <image.h>
+#include <log.h>
+#include <spl.h>
+#include <spl_load.h>
+#include <asm/io.h>
+#include <mtd.h>
+
+uint32_t __weak spl_mtd_get_uboot_offs(void)
+{
+	return CONFIG_SYS_MTD_U_BOOT_OFFS;
+}
+
+static ulong spl_mtd_read(struct spl_load_info *load, ulong offs, ulong size,
+			   void *dst)
+{
+	int err;
+	struct mtd_info *mtd = load->priv;
+	size_t ret_len;
+
+	debug("%s: offs %lx, size %lx, dst %p\n",
+	      __func__, offs, size, dst);
+
+	err = mtd_read(mtd, offs, size, &ret_len, dst);
+	if (err) {
+		printf("%s: error reading mtd, err - %d\n",
+		       __func__, err);
+		return 0;
+	}
+
+	if (ret_len != size) {
+		printf("%s: error reading mtd, ret_len != size\n",
+		       __func__);
+		return 0;
+	}
+
+	return size;
+}
+
+static int spl_mtd_load_image(struct spl_image_info *spl_image,
+			      struct spl_boot_device *bootdev)
+{
+	struct spl_load_info load;
+	struct mtd_info *mtd;
+
+	mtd_probe_devices();
+
+	switch (bootdev->boot_device) {
+	case BOOT_DEVICE_SPINAND:
+		mtd = get_mtd_device_nm("spi-nand0");
+		if (IS_ERR_OR_NULL(mtd)) {
+			printf("MTD device %s not found, ret %ld\n", "spi-nand",
+			       PTR_ERR(mtd));
+			return PTR_ERR(mtd);
+		}
+		break;
+	default:
+		puts("Unsupported MTD Boot Device!\n");
+		return -EINVAL;
+	}
+
+	spl_load_init(&load, spl_mtd_read, mtd, 1);
+	return spl_load(spl_image, bootdev, &load, 0, spl_mtd_get_uboot_offs());
+}
+
+SPL_LOAD_IMAGE_METHOD("SPINAND", 0, BOOT_DEVICE_SPINAND, spl_mtd_load_image);
diff --git a/drivers/mtd/Makefile b/drivers/mtd/Makefile
index ce05e206073..e62dafa1ba1 100644
--- a/drivers/mtd/Makefile
+++ b/drivers/mtd/Makefile
@@ -35,6 +35,7 @@ ifneq ($(mtd-y),)
 obj-$(CONFIG_SPL_MTD) += mtd.o
 endif
 obj-$(CONFIG_$(PHASE_)NAND_SUPPORT) += nand/
+obj-$(CONFIG_$(PHASE_)SPINAND_SUPPORT) += nand/
 obj-$(CONFIG_SPL_ONENAND_SUPPORT) += onenand/
 obj-$(CONFIG_$(PHASE_)SPI_FLASH_SUPPORT) += spi/
 obj-$(CONFIG_SPL_UBI) += ubispl/
diff --git a/drivers/mtd/mtd_uboot.c b/drivers/mtd/mtd_uboot.c
index 69cb3b51f92..f6402a59462 100644
--- a/drivers/mtd/mtd_uboot.c
+++ b/drivers/mtd/mtd_uboot.c
@@ -124,7 +124,7 @@ static void mtd_probe_uclass_spi_nor_devs(void)
 static void mtd_probe_uclass_spi_nor_devs(void) { }
 #endif
 
-#if defined(CONFIG_MTD_PARTITIONS)
+#if defined(CONFIG_MTD_PARTITIONS) && !defined(CONFIG_XPL_BUILD)
 
 #define MTDPARTS_MAXLEN         512
 
diff --git a/drivers/mtd/nand/Makefile b/drivers/mtd/nand/Makefile
index c8169cf7390..454af1819ba 100644
--- a/drivers/mtd/nand/Makefile
+++ b/drivers/mtd/nand/Makefile
@@ -6,5 +6,7 @@ obj-$(CONFIG_MTD_NAND_CORE) += nandcore.o
 obj-$(CONFIG_MTD_RAW_NAND) += raw/
 obj-$(CONFIG_MTD_SPI_NAND) += spi/
 else
+nandcore-objs := core.o bbt.o
 obj-$(CONFIG_$(PHASE_)NAND_SUPPORT) += raw/
+obj-$(CONFIG_$(PHASE_)SPINAND_SUPPORT) += spi/ nandcore.o
 endif
diff --git a/drivers/mtd/nand/spi/Makefile b/drivers/mtd/nand/spi/Makefile
index a7a0b2cb4b9..0f80f2671ea 100644
--- a/drivers/mtd/nand/spi/Makefile
+++ b/drivers/mtd/nand/spi/Makefile
@@ -3,4 +3,5 @@
 spinand-objs := core.o otp.o
 spinand-objs += alliancememory.o ato.o esmt.o fmsh.o foresee.o gigadevice.o macronix.o
 spinand-objs += micron.o paragon.o skyhigh.o toshiba.o winbond.o xtx.o
+obj-$(CONFIG_$(PHASE_)SPINAND_SUPPORT) += spinand.o
 obj-$(CONFIG_MTD_SPI_NAND) += spinand.o
